### PR TITLE
feat: Allow user to select the source for the metric timestamp.

### LINKED ISSUE #9011: Telegraf OPCUA input doesn't report OPCUA server timestamps
### Relevant telegraf.conf:

```[[inputs.opcua]]
#   ## Device name
name = "DelpharmWater"
#   #
#   ## OPC UA Endpoint URL
#HMI IP
endpoint = "opc.tcp://192.168.1.197:4870"
#   #
#   ## Maximum time allowed to establish a connect to the endpoint.
connect_timeout = "1s"
#   #
#   ## Maximum time allowed for a request over the estabilished connection.
request_timeout = "2s"
#   #
#   ## Security policy, one of "None", "Basic128Rsa15", "Basic256",
#   ## "Basic256Sha256", or "auto"
security_policy = "None"
#   #
#   ## Security mode, one of "None", "Sign", "SignAndEncrypt", or "auto"
security_mode = "None"
#   #
#   ## Path to cert.pem. Required when security mode or policy isn't "None".
#   ## If cert path is not supplied, self-signed cert and key will be generated.
#   # certificate = "/etc/telegraf/cert.pem"
#   #
#   ## Path to private key.pem. Required when security mode or policy isn't "None".
#   ## If key path is not supplied, self-signed cert and key will be generated.
#   # private_key = "/etc/telegraf/key.pem"
#   #
#   ## Authentication Method, one of "Certificate", "UserName", or "Anonymous".  To
#   ## authenticate using a specific ID, select 'Certificate' or 'UserName'
auth_method = "Anonymous"
#   #
#   ## Username. Required for auth_method = "UserName"
#   # username = ""
#   #
#   ## Password. Required for auth_method = "UserName"
#   # password = ""
#   #
#   ## Node ID configuration
#   ## name                             - the variable name
--
#   ## identifier                       - tag as shown in opcua browser
#   ## data_type                        - boolean, byte, short, int, uint, uint16, int16,
#   ##                        uint32, int32, float, double, string, datetime, number
#   ## Example:
#   ## {name="ProductUri", namespace="0", identifier_type="i", identifier="2262", data_type="string", description="http://open62541.org"}
   nodes = [
{name="GRAPHIQUE_AIT_056",namespace="3",identifier_type="s",identifier="GRAPHIQUE_AIT_056",data_type="float",description="104_E_ANALOG_REAL.AIT_056"},
{name="GRAPHIQUE_FIT_050",namespace="3",identifier_type="s",identifier="GRAPHIQUE_FIT_050",data_type="float",description="104_E_ANALOG_REAL.FIT_050"},
{name="GRAPHIQUE_FIT_175",namespace="3",identifier_type="s",identifier="GRAPHIQUE_FIT_175",data_type="float",description="104_E_ANALOG_REAL.FIT_175"},
{name="GRAPHIQUE_AIT_177",namespace="3",identifier_type="s",identifier="GRAPHIQUE_AIT_177",data_type="float",description="104_E_ANALOG_REAL.AIT_177"},
{name="GRAPHIQUE_TT_230",namespace="3",identifier_type="s",identifier="GRAPHIQUE_TT_230",data_type="float",description="104_E_ANALOG_REAL.TT_230"},
{name="GRAPHIQUE_PT_228",namespace="3",identifier_type="s",identifier="GRAPHIQUE_PT_228",data_type="float",description="104_E_ANALOG_REAL.PT_228"},
{name="GRAPHIQUE_COT_AIT227",namespace="3",identifier_type="s",identifier="GRAPHIQUE_COT_AIT227",data_type="float",description="104_E_ANALOG_REAL.COT_AIT227"},
{name="GRAPHIQUE_AIT_227",namespace="3",identifier_type="s",identifier="GRAPHIQUE_AIT_227",data_type="float",description="104_E_ANALOG_REAL.AIT_227"},
{name="GRAPHIQUE_PT_205",namespace="3",identifier_type="s",identifier="GRAPHIQUE_PT_205",data_type="float",description="104_E_ANALOG_REAL.PT_205"},
{name="GRAPHIQUE_PT_204",namespace="3",identifier_type="s",identifier="GRAPHIQUE_PT_204",data_type="float",description="104_E_ANALOG_REAL.PT_204"},
{name="GRAPHIQUE_AIT_166",namespace="3",identifier_type="s",identifier="GRAPHIQUE_AIT_166",data_type="float",description="104_E_ANALOG_REAL.AIT_166"},
{name="GRAPHIQUE_FIT_226",namespace="3",identifier_type="s",identifier="GRAPHIQUE_FIT_226",data_type="float",description="104_E_ANALOG_REAL.FIT_226"},
{name="GRAPHIQUE_AIT_245_XV229",namespace="3",identifier_type="s",identifier="GRAPHIQUE_AIT_245_XV229",data_type="float",description="005_DB_Com_IHM.AIT245.Val_XV229"},
{name="GRAPHIQUE_Cuve",namespace="3",identifier_type="s",identifier="GRAPHIQUE_Cuve",data_type="float",description="005_DB_Com_IHM.LT_R200_Val"},
{name="GRAPHIQUE_AIT_245_XV239",namespace="3",identifier_type="s",identifier="GRAPHIQUE_AIT_245_XV239",data_type="float",description="005_DB_Com_IHM.AIT245.Val_XV239"},
{name="GRAPHIQUE_AIT_245_XV240",namespace="3",identifier_type="s",identifier="GRAPHIQUE_AIT_245_XV240",data_type="float",description="005_DB_Com_IHM.AIT245.Val_XV240"},
]  


```

### Steps to reproduce:

1. set the time to your OPCUA different than the one where Telegraf is running
2. Collect OPCUA nodes using telegraf with similar config as the one provided 
3. configure Telegraf to output to influxDB or to a file with influx format 
4. compare the timestamps exposed by Telegraf on the output 

### Expected behavior:

The timestamps should be matching the OPCUA server timestamps and not the telegraf server 's local time

### Actual behavior:

Actually the timestamps is not the one provided by the OPCUA server but the local one , which is not the expected behaviour
### Additional info:
Server opcua Timestamp is well collected by the OPCUA input plugin ( func get data ) but not exposed by gather function

### PR DESCRIPTION
**Pull request by srebhan**
### Required for all PRs:

- [X] Associated README.md updated.
- [ ] Has appropriate unit tests.

Closes #9011.

### CODE REVIEW

**Comment by jordanbauduin** [comment]
Test done on PLC Siemens OPCUA server , timestamp is correctly taken from the node when source or server are selected 
thx for fixing it so fast !

File: plugins/inputs/opcua/README.md [hunk 1]
```diff
@@ -46,6 +46,12 @@ Plugin minimum tested version: 1.16
   ## Password. Required for auth_method = "UserName"
   # password = ""
   #
+  ## Option to select the metric timestamp to use. Valid options are:
+	##     "gather" -- uses the time of receiving the data in telegraf
+	##     "server" -- uses the timestamp provided by the server
```
Comment (ssoroka) [inline plugins/inputs/opcua/README.md:51]: shouldn't gather and server be the same thing?

**Review by ssoroka — CHANGES_REQUESTED** [review]
this is a common ask across a lot of plugins. I kind of feel like it shouldn't go into a single plugin. I was considering making a plugin for it, but with the new Starlark time support, I'm not sure it's needed.

**Comment by jordanbauduin** [comment]
I there, 

 

In OPC  UA you can have 2 topologies ( for most ) 

 

First is the client ( telegraf ) connecting to the OPCUA server 

 

*	In such case the source et server  timestamp are the same.

 

Second topology is having many servers sending to one server acting as proxy 

 

*	In this case the source timestamp might be different from the server ( proxy ) 

 

This is available here => https://reference.opcfoundation.org/v104/Core/docs/Part4/3.1/

 

The last option we have added is the gather , which in case could take the timestamp from the client ( telegraf ) 

 

I hope it helps 

 

Kr 

 

Jordan 

 

De : Steven Soroka ***@***.***> 
Envoyé : jeudi 18 mars 2021 20:03
À : influxdata/telegraf ***@***.***>
Cc : jordan bauduin ***@***.***Com>; Comment ***@***.***>
Objet : Re: [influxdata/telegraf] Allow user to select the source for the metric timestamp. (#9013)

 

@ssoroka requested changes on this pull request.

this is a common ask across a lot of plugins. I kind of feel like it shouldn't go into a single plugin. I was considering making a plugin for it, but with the new Starlark time support, I'm not sure it's needed.

  _____  

In plugins/inputs/opcua/README.md <https://github.com/influxdata/telegraf/pull/9013#discussion_r597160584> :

> +     ##     "gather" -- uses the time of receiving the data in telegraf
+       ##     "server" -- uses the timestamp provided by the server

shouldn't gather and server be the same thing?

**Review by Jagularr — APPROVED** [review]

**Comment by reimda** [comment]
@jordanbauduin Is a single plugin-level setting sufficient?  I wonder if people will want some nodes (or groups) to use one timestamp and other nodes/groups to use a different timestamp.

For example suppose you have many servers feeding a proxy server.  You might want to use the timestamp from the feeding server if it's accurate, but if the feeding server's timestamp isn't accurate (maybe no real time clock or ntpd) you would want the timestamp from the proxy server.

Is this a useful design or more flexibility than anyone would ever need?

**Comment by srebhan** [comment]
@jordanbauduin can you please comment on the question above.

File: plugins/inputs/opcua/README.md (hunk 1 above)
Comment (Hipska) [reply plugins/inputs/opcua/README.md:51]: "gather" is what is the current behaviour I think..

**Review by Hipska — APPROVED** [review]
Approved, there is only a small linter check of an empty line that can be fixed..

**Comment by srebhan** [comment]
@Hipska fixed. Thanks!

**Comment by reimda** [comment]
I'm ok taking this PR as it is. I still wonder if people will need more flexibility and want to choose the source for each group or node, but that shouldn't prevent us from starting with a single setting at the plugin level. If users ask for more flexibility we can revisit adding a timestamp setting to groups or nodes.

**Comment by reimda** [comment]
Sven, could you fix the conflicts and we'll merge this?

**Comment by srebhan** [comment]
Hey @reimda sorry for taking so long. Rebased and title fixed.

**Pull request merged**